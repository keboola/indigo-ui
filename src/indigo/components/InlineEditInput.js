import React from 'react';
import PropTypes from 'prop-types';
import { Button, FormControl, OverlayTrigger, Tooltip } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons/faTimesCircle';
import { faPencilAlt } from '@fortawesome/free-solid-svg-icons/faPencilAlt';
import Loader from './Loader';

class InlineEditTextInput extends React.Component {
  constructor(props) {
    super(props);
    this.handleOnSubmit = this.handleOnSubmit.bind(this);
    this.handleOnChange = this.handleOnChange.bind(this);
  }

  render() {
    if (this.props.isEditing) {
      return this.renderEditInput();
    }

    return this.renderStaticInput();
  }

  handleOnChange(e) {
    return this.props.onEditChange(e.target.value);
  }

  handleOnSubmit(e) {
    e.preventDefault();
    return this.props.onEditSubmit();
  }

  renderEditInput() {
    return (
      <div className="inline-edit-input">
        <form className="inline-edit-input-form" onSubmit={this.handleOnSubmit}>
          <FormControl
            className="inline-edit-input-control"
            type="text"
            value={this.props.text}
            disabled={this.props.isSaving}
            placeholder={this.props.placeholder}
            onChange={this.handleOnChange}
            autoFocus
          />

          <div className="inline-edit-input-buttons">
            <Button
              className="inline-edit-input-cancel"
              bsStyle="link"
              disabled={this.props.isSaving}
              onClick={this.props.onEditCancel}
            >
              <FontAwesomeIcon icon={faTimesCircle} />
            </Button>
            <Button
              className="inline-edit-input-submit"
              bsStyle="primary"
              disabled={this.props.isSaving || !this.props.isValid}
              type="submit"
            >
              Save
            </Button>
            {this.props.isSaving && <Loader />}
          </div>
        </form>
      </div>
    );
  }

  renderStaticInput() {
    const tooltip = <Tooltip id="inline-edit-input-tooltip">{this.props.editTooltip}</Tooltip>;

    return (
      <OverlayTrigger placement={this.props.tooltipPlacement} overlay={tooltip}>
        <span className="inline-edit-input-link" onClick={this.props.onEditStart}>
          {this.props.text ? (
            <span>{this.props.text}</span>
          ) : (
            <span className="text-muted">{this.props.placeholder}</span>
          )}
          <FontAwesomeIcon icon={faPencilAlt} fixedWidth className="icon-addon-left" />
        </span>
      </OverlayTrigger>
    );
  }
}

InlineEditTextInput.propTypes = {
  onEditStart: PropTypes.func.isRequired,
  onEditCancel: PropTypes.func.isRequired,
  onEditChange: PropTypes.func.isRequired,
  onEditSubmit: PropTypes.func.isRequired,
  text: PropTypes.string,
  isSaving: PropTypes.bool,
  isEditing: PropTypes.bool,
  isValid: PropTypes.bool,
  editTooltip: PropTypes.string,
  tooltipPlacement: PropTypes.string,
  placeholder: PropTypes.string,
};

InlineEditTextInput.defaultProps = {
  placeholder: 'Click to edit',
  editTooltip: 'Click to edit',
  tooltipPlacement: 'top',
  isSaving: false,
  isEditing: false,
};

export default InlineEditTextInput;
