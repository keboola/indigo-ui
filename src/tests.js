import renderer from 'react-test-renderer';

export const snapshot = (component) => {
  const tree = renderer
    .create(component)
    .toJSON();
  expect(tree).toMatchSnapshot();
};
