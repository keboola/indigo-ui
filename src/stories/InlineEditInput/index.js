import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import InlineEditInput from '../../indigo/components/InlineEditInput';

export default class InlineEditInputExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      initialText: '',
      text: '',
      isEditing: false,
      isSaving: false,
    };
  }

  render() {
    return (
      <InlineEditInput
        isEditing={this.state.isEditing}
        isValid
        onEditStart={() => {
          action('editing started')();
          this.setState({
            isEditing: true,
          });
        }}
        onEditCancel={() => {
          action('editing cancelled')();
          this.setState({
            isEditing: false,
            text: this.state.initialText,
          });
        }}
        onEditChange={(text) => {
          this.setState({
            text,
          });
        }}
        onEditSubmit={() => {
          action('form submitted')();
          this.setState({
            isSaving: true,
          });
          setTimeout(() => {
            this.setState({
              initialText: this.state.text,
              isEditing: false,
              isSaving: false,
            });
          }, 1000);
        }}
        text={this.state.text}
        isSaving={this.state.isSaving}
        editTooltip="Click to edit"
        placeholder="Click to edit"
      />
    );
  }
}

storiesOf('InlineEditInput', module)
  .add(
    'Read mode',
    () => {
      return (
        <InlineEditInput
          isEditing={false}
          isValid
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    },
    { info: { text: 'Input in read only.' } }
  )
  .add(
    'Edit mode',
    () => {
      return (
        <InlineEditInput
          isEditing
          isValid
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    },
    { info: { text: 'Input in editing mode.' } }
  )
  .add(
    'Edit mode - saving',
    () => {
      return (
        <InlineEditInput
          isEditing
          isValid
          isSaving
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    },
    { info: { text: 'Input in editing mode.' } }
  )
  .add(
    'Working example',
    () => {
      return (
        <div>
          some text <InlineEditInputExample />
        </div>
      );
    },
    { info: { text: 'Full example with actions, etc.' } }
  );
