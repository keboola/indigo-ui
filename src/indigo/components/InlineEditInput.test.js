import React from 'react';

import { snapshot } from '../../tests';
import InlineEditInput from './InlineEditInput';

describe('<InlineEditInput />', () => {
  const compulsoryProps = {
    onEditStart: () => null,
    onEditCancel: () => null,
    onEditChange: () => null,
    onEditSubmit: () => null,
  };

  it('InlineEditInput - default', () => {
    snapshot(<InlineEditInput {...compulsoryProps} />);
  });

  it('InlineEditInput - read mode', () => {
    snapshot(<InlineEditInput {...compulsoryProps} isEditing={false} />);
  });

  it('InlineEditInput - edit mode', () => {
    snapshot(<InlineEditInput {...compulsoryProps} isEditing />);
  });

  it('InlineEditInput - saving in edit mode', () => {
    snapshot(<InlineEditInput {...compulsoryProps} isEditing isSaving />);
  });

  it('InlineEditInput - disabled save button when text is not valid', () => {
    snapshot(<InlineEditInput {...compulsoryProps} isEditing isValid={false} />);
  });
});
