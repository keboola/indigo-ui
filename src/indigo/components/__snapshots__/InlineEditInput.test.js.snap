// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineEditInput /> InlineEditInput - default 1`] = `
<span
  className="inline-edit-input-link"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="text-muted"
  >
    Click to edit
  </span>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-edit fa-w-18 icon-fixed-width icon-addon-left"
    data-icon="edit"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</span>
`;

exports[`<InlineEditInput /> InlineEditInput - disabled save button when text is not valid 1`] = `
<div
  className="inline-edit-input"
>
  <form
    className="inline-edit-input-form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="inline-edit-input-control form-control"
      disabled={false}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
    />
    <div
      className="inline-edit-input-buttons"
    >
      <button
        className="inline-edit-input-cancel btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 "
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="inline-edit-input-submit btn btn-sm btn-primary"
        disabled={true}
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;

exports[`<InlineEditInput /> InlineEditInput - edit mode 1`] = `
<div
  className="inline-edit-input"
>
  <form
    className="inline-edit-input-form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="inline-edit-input-control form-control"
      disabled={false}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
    />
    <div
      className="inline-edit-input-buttons"
    >
      <button
        className="inline-edit-input-cancel btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 "
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="inline-edit-input-submit btn btn-sm btn-primary"
        disabled={true}
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;

exports[`<InlineEditInput /> InlineEditInput - read mode 1`] = `
<span
  className="inline-edit-input-link"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="text-muted"
  >
    Click to edit
  </span>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-edit fa-w-18 icon-fixed-width icon-addon-left"
    data-icon="edit"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</span>
`;

exports[`<InlineEditInput /> InlineEditInput - saving in edit mode 1`] = `
<div
  className="inline-edit-input"
>
  <form
    className="inline-edit-input-form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="inline-edit-input-control form-control"
      disabled={true}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
    />
    <div
      className="inline-edit-input-buttons"
    >
      <button
        className="inline-edit-input-cancel btn btn-link"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 "
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="inline-edit-input-submit btn btn-sm btn-primary"
        disabled={true}
        type="submit"
      >
        Save
      </button>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-spinner fa-w-16 fa-spin icon-fixed-width"
        data-icon="spinner"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </form>
</div>
`;
